Add ERD

a0ki-github opened the pull request:
READMEにER図の画像とURLリンクを追加。

[![Image from Gyazo](https://i.gyazo.com/d6395652606778ef60d89164dcfd3f79.png)](https://drive.google.com/file/d/1-CTAXc4nQxtcXVW81UpkotjkiVykD2OA/view?usp=sharing)

## 特にレビューいただきたい点
下記のテーブル設計の考え方に不足や誤りがないか、レビューいただけるとありがたいです。

### 考え方：
ユーザー同士がそれぞれの得点を競うゲームなので、
例えば管理者の操作ミスなど、何か事故が起こった際のトレーサビリティを重視しています。

そのためには、ポイントの減点処理は管理者によるコンソール操作やAdminコントローラを通して行われるのではなく、
ユーザーのBET履歴と候補者の脱落履歴をもとにポイントが計算される仕組みが必要と考えています。
（わざわざ中間テーブル「Bettings」を作るのはそのためです。）

ポイントに関するテーブルやカラムがないのもそのためであり、
この場合、ポイント計算は以下のようにapp/model/user.rb内のモデルメソッドにて行うのが良いと考えています。

```
def points
  points = 100

  # 第一回放送時点でBETを外している場合に減点する処理
    # この時点より前に作られた最新BETと、この放送回で脱落した候補者を参照し、一致した場合は減点
    # 誰にもBETしていない場合も同様に減点

  # 第二回放送時点でBETを外している場合に減点する処理
    # この時点より前に作られた最新BETと、この放送回以前に脱落した候補者を参照し、一致した場合は減点
    # 誰にもBETしていない場合も同様に減点

  # 第三回放送時点でBETを外している場合に減点する処理
    # この時点より前に作られた最新BETと、この放送回以前に脱落した候補者を参照し、一致した場合は減点
    # 誰にもBETしていない場合も同様に減点

  # 最終回放送時点でBETを外している場合に減点する処理
    # この時点より前に作られた最新BETと、この放送回以前に脱落した候補者を参照し、一致した場合は減点
    # 誰にもBETしていない場合も同様に減点

  # BETの変更回数に応じて減点する処理

  return points #　ユーザーの最終点数
end
```

## 補足
### 番組ルール：
登場人物は男性1人と女性15人。男性は女性が1人になるまで脱落させていく。放送は全4回。

### アプリルール：
・ユーザーは登録時に持ち点として100ポイント付与
・ユーザーはバチェラー参加者1名のみBET可能で、以下の場合に持ち点を減点
　・番組開始後にBETするメンバーを変更すると1回につき-10ポイント
　・各放送日時点で脱落メンバーにBETしているユーザー、\
　　または誰にもBETしていないユーザーには-20ポイント
・番組終了時に高得点上位何名かを発表

お手数ですが、ご確認のほどよろしくお願いいたします。

a0ki-github committed:
Fix: 旧タイトルを修正

a0ki-github committed:
Add: READMEにER図を追加

a0ki-github committed:
Add: 最終行追加

a0ki-github committed:
Fix: ルールを分かりやすく書き直し

a0ki-github committed:
Fix: ルールを分かりやすく書き直し

MrFuku commented:
## users
- passwordカラムはそうならないかと思います

## bettings
- railsであればcreated_atはデフォルトで作成されます。他のテーブルでは省略していたので、全体として省略するか省略しないか統一したほうが良いと思います

## その他
- 放送日（脱落者が決まる日）はどういった形で持ちますか？事前に決まっているのであればコードに書いておくでも良いですが、一般的にはDBに保存するのが良い気がします（その場合はcandidatesとの正規化を考える必要がありそうです）。

a0ki-github committed:
Fix: 以下理由によりER図を修正
・Usersテーブルのpasswordカラムがsorceryを利用した場合の実態に即していなかった
・Bettingsテーブルに省略されているはずのcreated_atカラムが記述されていた
・各候補者の脱落日をCandidatesテーブルのdropout_dateカラムで管理する方法に変更した

a0ki-github commented:
> ## users
> * passwordカラムはそうならないかと思います

ご指摘ありがとうございます。
今回はsorceryを使用予定ですので、正しくは
```
crypted_password str
salt str
```
でした。


> ## bettings
> * railsであればcreated_atはデフォルトで作成されます。他のテーブルでは省略していたので、全体として省略するか省略しないか統一したほうが良いと思います

すみません、確かに分かりづらくなっておりました。。
他のテーブルに合わせ省略しました！
 
> ## その他
> * 放送日（脱落者が決まる日）はどういった形で持ちますか？事前に決まっているのであればコードに書いておくでも良いですが、一般的にはDBに保存するのが良い気がします（その場合はcandidatesとの正規化を考える必要がありそうです）。

放送日は11/25、12/02、12/09、12/16の4つ（いずれも22:00）と決められているのですが、
確かに自作の変数に格納するよりはDBを利用していた方が可読性も高いと思いました。

一方でidで管理するほどのものでもなさそうに感じたので、Candidatesテーブル内に「脱落が決まる日」として
dropout_dateカラムを設置し、以下のようにenumを用いて列挙するのが良いと考えました。
```
# app/model/candidate.rb
class Candidate < ApplicationRecord
  enum dropout_date: { first_dropout: Time.new(2021, 11, 25, 22),
                       second_dropout: Time.new(2021, 12, 2, 22),
                       third_dropout: Time.new(2021, 12, 9, 22),
                       fourth_dropout: Time.new(2021, 12, 16, 22) }
end
```
これにより、各放送日はCandidate.dropout_dates[:first_dropout]などで取得でき、
一番使いそうな「ある放送日に脱落した候補者の集合」はCandidates.first_dropoutで取得できます。

またenumを使う場合、もともと考えていたstatusカラムは内容が重複するので削除しました。
（生き残っていることはdropout_date = nil で表現可能）

#### 補足
DropoutDatesテーブル（自作カラムはdropout_dateのみ）を新設すれば以下の案も採用できるかと考えましたが、
検討の結果enumがベストと考えました。

①`DropoutDate has_many Candidates`のアソシエーションを定義する。
　→Candidatesが親であるDropoutDateを必須としてしまうので、生き残っている候補者が存在できなくなってしまう。
　　（生き残っている候補者用に9999年などに設定したDropoutDateインスタンスを作る手もあるとは思いますが、、）
②`Candidate has_one DropoutDate`のアソシエーションを定義する。
　→①の問題は解消されるが、一番使いそうな「ある放送日に脱落した候補者の集合」が
　　enumに比べると簡潔に取得できない
　
長文になりましたが、ご確認のほどよろしくお願いいたします。

murata0705 commented:
enum って、value に datetime オブジェクト渡せるんでしたっけ？

渡せるのであれば良いとは要件は満たせると思いますが、放送回と放送日を定数でコードに持って、DBのカラムには放送回を int で持たせるほうがシンプルかなと思います。（テーブルを作らないのであれば）

あと、datetime 型のカラムで _date は分かりづらいので、_datetime にするか、_at など、Rails の命名規則に合わせた方が良いかなと思います。

a0ki-github committed:
Fix: ER図を修正（候補者の脱落情報を「いつ脱落したか」ではなく「どの放送回まで出演していたか」によって管理する方法に変更）

a0ki-github commented:
> enum って、value に datetime オブジェクト渡せるんでしたっけ？
> 

手元で試したところ、渡せるようでした！

> 渡せるのであれば良いとは要件は満たせると思いますが、放送回と放送日を定数でコードに持って、DBのカラムには放送回を int で持たせるほうがシンプルかなと思います。（テーブルを作らないのであれば）
> 

**前言撤回で申し訳ないのですが、やはりテーブルを作った方がよいのではないかと考えました。**
私の最新案は以下となります。（赤枠が変更部分です）
[![Image from Gyazo](https://i.gyazo.com/4842dfcef4701cf1572aeb68e55b6956.png)](https://gyazo.com/4842dfcef4701cf1572aeb68e55b6956)

ポイント
・候補者の脱落情報を、「いつ脱落したか」ではなく「どの放送回まで出演していたか」によって管理する方法に変更。
・候補者は自身が出演している放送回を所有。逆に放送回はそこに出演している候補者を所有する。
・アソシエーションについては、双方に`has_and_belongs_to_many`を定義。

こちらの方がenumを活用する方法よりも欲しいデータの呼び出しが自然であり、
一番使いそうな「ある放送日に脱落した候補者の集合」も、少し複雑かもしれないですがscopeを活用すれば
簡潔に呼び出せると思います。

お手数ですが、再度レビューのほどよろしくお願いいたします。

murata0705 commented:
放送日が終わるごとに次の放送日分の`candidate_episodes`レコードを管理側で作るイメージですかね？（その会で脱落した人を除く）

大丈夫だと思います。

a0ki-github commented:
> 放送日が終わるごとに次の放送日分の`candidate_episodes`レコードを管理側で作るイメージですかね？（その会で脱落した人を除く）
>
そのイメージです！
 
> 大丈夫だと思います。

ありがとうございます、実装に進みます！！
